lib/auth/gitserver/gitserverv1/service.go

```diff
@@ -0,0 +1,185 @@
+ */
+
+package gitserverv1
+
+import (
+	"context"
+	"log/slog"
+
+	"github.com/gravitational/trace"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/gravitational/teleport"
+	pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/gitserver/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/authz"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// Config is the config for Service.
+type Config struct {
+	// Authorizer is the authorizer to use.
+	Authorizer authz.Authorizer
+	// Backend is the backend for storing UserTask.
+	Backend services.GitServers
+	// Log is the slog logger.
+	Log *slog.Logger
+}
+
+// Service implements the gRPC service that manages git servers.
+type Service struct {
+	pb.UnsafeGitServerServiceServer
+
+	cfg Config
+}
+
+// NewService creates a new git server service.
+func NewService(cfg Config) (*Service, error) {
+	if cfg.Authorizer == nil {
+		return nil, trace.BadParameter("authorizer is required")
+	}
+	if cfg.Backend == nil {
+		return nil, trace.BadParameter("backend is required")
+	}
+	if cfg.Log == nil {
+		cfg.Log = slog.With(teleport.ComponentKey, "gitserver.service")
+	}
+	return &Service{
+		cfg: cfg,
+	}, nil
+
+}
+
+func toServerV2(server types.Server) (*types.ServerV2, error) {
+	serverV2, ok := server.(*types.ServerV2)
+	if !ok {
+		return nil, trace.Errorf("encountered unexpected server type: %T", serverV2)
+	}
+	return serverV2, nil
+}
+
+func (s *Service) CreateGitServer(ctx context.Context, req *pb.CreateGitServerRequest) (*types.ServerV2, error) {
+	if _, err := s.authorize(ctx, types.VerbCreate); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	server, err := s.cfg.Backend.CreateGitServer(ctx, req.Server)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return toServerV2(server)
+}
+
+func (s *Service) GetGitServer(ctx context.Context, req *pb.GetGitServerRequest) (*types.ServerV2, error) {
+	authCtx, err := s.authorize(ctx, types.VerbRead)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	server, err := s.cfg.Backend.GetGitServer(ctx, req.Name)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := s.checkAccess(authCtx, server); err != nil {
+		if trace.IsAccessDenied(err) {
+			return nil, trace.NotFound("not found")
+		}
+		return nil, trace.Wrap(err)
+	}
+	return toServerV2(server)
+}
+
+func (s *Service) ListGitServers(ctx context.Context, req *pb.ListGitServersRequest) (*pb.ListGitServersResponse, error) {
+	authCtx, err := s.authorize(ctx, types.VerbRead, types.VerbList)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	servers, next, err := s.cfg.Backend.ListGitServers(ctx, int(req.PageSize), req.PageToken)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	resp := &pb.ListGitServersResponse{
+		NextPageToken: next,
+	}
+	for _, server := range servers {
+		err := s.checkAccess(authCtx, server)
+		if trace.IsAccessDenied(err) {
```

Comment:
In the current version  if a user has access to the GitHub server via labels but does not have access to `checkAccessToKind(types.KindGitServer, verb, additionalVerbs...)` the `ListGitServers function` will return “access denied.”

The question is: 
Why does a user need to have both `checkAccessToKind(types.KindGitServer, verb, additionalVerbs...)` and `s.checkAccess(authCtx, server)` permissions to list GitHub servers when `types.KindGitServer`  access already provides access to list and read all GitHub servers?


I assume the logic should allow listing servers if a user has read or list access to the types.KindGitServer resource kind otherwise the types.KindGitServer verbReadList don't make sens because additional permission are needed. 

I think that checkAccess  should be used  as a fallback, and the condition used  between `s.authorize` and `s.checkAccess`  should be  OR operator, not AND.

Reply:
just following the pattern for existing heartbeat resources like `GetNodes`, `GetApplicationServers`.

But I found the issue with the new resource. We use this to make sure users without exiplict rules can access the heartbeats:
https://github.com/gravitational/teleport/blob/6f73d959e1c1725fb97bb00b95fc24d2133f3b60/lib/services/role.go#L59

i will add RO there for git_server and update UT accordingly.

Reply:
https://github.com/gravitational/teleport/pull/49393/commits/e3562bc7e23eebc8e6c14fd0e7addd1ba9e1d931